### PR TITLE
MMB-240: Add new field to choose contact for contribution

### PR DESCRIPTION
**Pull request by viraj-compuco**
Overview
----------------------------------------
This PR aims to add new field to contribution section so that user specify which contact will be the contribution contact (i.e. the invoice will be addressed to).

Before
----------------------------------------
Default contact 1 is always passed.

After
----------------------------------------
Now user can have field to choose the required contact for contribution

Screenshot 2023-11-17 at 5 21 10 PM

**Error message**
Screenshot 2023-12-04 at 7 22 36 PM

**Error message for User select value**
Screenshot 2023-12-06 at 6 33 59 PM

Technical Details
----------------------------------------
- Added Contribution Contact select list field to `wf_crm_get_fields`  method under `webform_civicrm/includes/utils.inc` file.
- Have data type for this new field as `ContactReference`, It will populate the list of contact selected.
- Added a logic to get the field value and pass it to contact ID param in `contributionParams()` method under the file - `webform_civicrm/includes/wf_crm_webform_postprocess.inc`.
- Have added the JS changes in `webform_civicrm/js/webform_civicrm_admin.js` to validate the email field enabled for other contacts.
- Created new function `billingMessagesUserSelect()` to handle the multiple error messages when Contribution contact field value is selected as `User select`
- Added new email param in `createContact` method to avoid duplicate contact creation.

### CODE REVIEW

File: includes/wf_crm_webform_postprocess.inc [hunk 1]
```diff
@@ -2035,7 +2035,7 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
     $params['financial_type_id'] = $this->contribution_page['financial_type_id'];
     $params['currency'] = $params['currencyID'] = $this->contribution_page['currency'];
     $params['skipRecentView'] = $params['skipLineItem'] = 1;
-    $params['contact_id'] = $this->ent['contact'][1]['id'];
+    $params['contact_id'] = $this->getData('civicrm_1_contribution_1_contribution_contact_id') ?? $this->ent['contact'][1]['id'];
```
Comment (omarabuhussein) [inline includes/wf_crm_webform_postprocess.inc]: You probably need to change this place as well:

https://github.com/compucorp/webform_civicrm/blob/ee465906b66a0b87827c69beb066bad6fd15feda/includes/wf_crm_webform_postprocess.inc#L1967

where I think this line:

```
    $contact = wf_civicrm_api('contact', 'getsingle', ['id' => $this->ent['contact'][1]['id']]);
```

should be replaced by:

```
$contact = wf_civicrm_api('contact', 'getsingle', ['id' => $params['contact_id']]);
```

**Review by omarabuhussein — CHANGES_REQUESTED** [review]

File: includes/wf_crm_webform_postprocess.inc (hunk 1 above)
Comment (viraj-compuco) [reply includes/wf_crm_webform_postprocess.inc]: Updated @omarabuhussein

**Review by omarabuhussein — APPROVED** [review]

**Review by omarabuhussein — APPROVED** [review]

File: includes/utils.inc [hunk 2]
```diff
@@ -1104,6 +1104,14 @@ function wf_crm_get_fields($var = 'fields') {
         'name' => t('Contribution Source'),
         'type' => 'textfield',
       ];
+      // Custom field for contribution contact.
+      $fields['contribution_contribution_contact_id'] = [
```
Comment (viraj-compuco) [inline includes/utils.inc:1107]: @omarabuhussein Here I updated the machine name for new field as it was conflicting with `contact_id` key.
Have added the contribution as prefix so in params we are getting the field like this `contribution_contact_id`

File: includes/utils.inc [hunk 3]
```diff
@@ -1104,6 +1104,14 @@ function wf_crm_get_fields($var = 'fields') {
         'name' => t('Contribution Source'),
         'type' => 'textfield',
       ];
+      // Custom field for contribution contact.
```
Comment (omarabuhussein) [inline includes/utils.inc]: "Custom field" has a specific meaning in CiviCRM, so using it like this might be confusing, I think you can just remove this line comment.

**Review by omarabuhussein — CHANGES_REQUESTED** [review]

File: includes/utils.inc (hunk 3 above)
Comment (viraj-compuco) [reply includes/utils.inc]: Removed

**Review by omarabuhussein — APPROVED** [review]

File: includes/wf_crm_webform_postprocess.inc [hunk 4]
```diff
@@ -627,6 +627,9 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
     $params = $contact['contact'][1];
     // CiviCRM API is too picky about this, imho
     $params['contact_type'] = ucfirst($params['contact_type']);
+    if (!empty($contact['email'])) {
```
Comment (viraj-compuco) [inline includes/wf_crm_webform_postprocess.inc]: @omarabuhussein I have added new `email` param in `createContact` method to avoid duplicate contact creation.
Here is the one scenario capture by Varun - When user try to create the new contact for existing contact field then 2 user gets create one **with** email address and one **without** email address which was not capture in `findDuplicateContact` method.
So I have added the email param when creating new contact and now the existing contact will be capture `findDuplicateContact` method.

Comment (jamienovick) [reply includes/wf_crm_webform_postprocess.inc]: Do we need to take into account Contact matching rules (dedupe rules) which can be configured per form? Or maybe that doesn't apply in this scenario? @viraj-compuco

Comment (viraj-compuco) [reply includes/wf_crm_webform_postprocess.inc]: @jamienovick I don't think so that would be required in this scenario.

File: includes/wf_crm_webform_postprocess.inc [hunk 5]
```diff
@@ -2035,7 +2039,29 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
...
+      $contribution_contact_cid = $contribution_contact_id;
+    }
+    else {
+      $contribution_contact_cid = $this->ent['contact'][$contribution_contact_id]['id'];
+    }
+    // For new contact cid is empty.
+    // Create new contact to get the cid.
+    if (!$contribution_contact_cid) {
+      $contact = $this->data['contact'][$contribution_contact_id];
+      $this->ent['contact'][$contribution_contact_id]['id'] = $contribution_contact_cid = $this->createContact($contact);
```
Comment (omarabuhussein) [inline includes/wf_crm_webform_postprocess.inc]: @viraj-compuco 

The duplicate contact that you are talking about above seems to be a ramification of you adding this new line, which is not a line that makes sense to be added here because this function purpose is to set the contribution parameters not to create contacts. Also all contacts at this point  (as far as I can see, but maybe I am missing something here) should already be created by the preSave() function so I am not sure why you have to do this here.  So unless you have a better solution than this  I will have to deploy this PR on my local and check few things before I can tell you if I can approve your new changes or not.

**Review by omarabuhussein — CHANGES_REQUESTED** [review]

File: includes/wf_crm_webform_postprocess.inc (hunk 5 above)
Comment (viraj-compuco) [reply includes/wf_crm_webform_postprocess.inc]: @omarabuhussein Lets have short call to discuss this changes

Comment (viraj-compuco) [reply includes/wf_crm_webform_postprocess.inc]: @jamienovick Quick question
We have to fill Billing Name and Address on the payment page, currently it is of Contact 1 only.
Now with new implementation, we are assigning the contribution to other contacts also, Can we change the Billing Name and Address for the selected contribution contact or keep it as it is for contact 1?

cc @omarabuhussein

Comment (jamienovick) [reply includes/wf_crm_webform_postprocess.inc]: The billing name and address should be attached to the contact who makes the payment. Thanks

Comment (omarabuhussein) [reply includes/wf_crm_webform_postprocess.inc]: @jamienovick 

Is it safe to say then that if you assign the contribution to another contact other than contact 1 (contact 1 is usually who submits the webform) then the contribution should be configured to be "Pay later" (Pending), given if a payment processor is used such Stripe , then Contact 1 will be the one that does both submitting the form and making the payment, so it does not make sense for the contribution to be assigned to another contact (unless you are going to assume that Contact 1 will use Contact 2 credit card or Stripe account to make the payment).

Comment (jamienovick) [reply includes/wf_crm_webform_postprocess.inc]: I actually don't think we care who the contact who submits the form is, even if its a payment processor. Let me run a scenario for you:

Eg: John Smith comes to form but wants to pay for membership for his organisation:

- Individual is Contact 1
- Org Contact is Contact 2
- Membership 1 is attached to Contact 2
- Contribution contact is Contact 2 (i.e. the Organisation)

We want the contribution (invoice) to be connected to contact 2. 
If paying by Stripe, John will enter his credit card details, but he is still paying for Org Contact 
The contribution will show the "billing" information as a separate non editable block at the bottom of the page:

image

I don't think there is any issue with this being a Stripe transaction or Pay later.

Basically we should think of the contribution as the "invoice" and the payment as a separate entity. The invoice would be addressed to contact 2, but the payment was made by contact 1. What would probably be preferable would be that the billing address information was a property of the payment, and not the contribution - I'm not sure what the data structure is for this but at least CiviCRM stores the details of who made the payment.

**Review by omarabuhussein — APPROVED** [review]

**Review by omarabuhussein — APPROVED** [review]
